### DISCUSSION TITLE
Update specific item in a list in the cache from subscription

### DISCUSSION THREAD

**Discussion by rasmusolssony** [urql-graphql/urql#3296] [Q&A]
Hi!

I have a query with lists of items, there are a few static elements to these items which is fetched with a normal query. There are also a few elements which change over time. Imagine that you draw a card over a reader and three fields update. Now I want to be able to subscribe to any number of these fields and then update the cache with this information.

Right now I use writeFragment but if I don't subscribe to all of the possible fileds it complains that those fields are undefined.

Is there any way to update fields of an item in a list similar to writeFragment without having all of the potential fields of that item present?

**Comment by kitten** [comment]
In the best case scenario, you'd be using `writeFragment` very sparingly.
So, this is often more a question of schema design and making your schema work for your UI.

For instance,
say you have an object:

```graphql
type Starship {
  id: ID!
  currentDistanceFromEarth: Float
}
```

And a subscription like this:

```graphql
type Subscription {
  starshipDistanceFromEarth(id: ID!): Float
}
```

This would be bad for you and your UI, because you now have to use `writeFragment` to hook the subscription up to your `Starship` type, by updating the field manually.

The same often goes for your mutations.
You can write a schema that simply doesn't give you enough information.

This is also often a problem with modeling relations and related types in queries. This could be solved with edges, i.e. types that represent shared relationships between objects.

In other words, say your subscription now looks like this:

```graphql
type Subscription {
  starshipDistanceFromEarth(id: ID!): Starship!
  starshipUpdate(id: ID!): Starship!
}
```

In either case, above, if you now have a query and a subscription, there's no added cache update code you would need to write.

In general,
- the less you have to use `writeFragment` for value updates, the better
- sometimes of course, you can't get around it, but these cases are explicitly rarer and often should only occur with side-effects
- optimally, your updaters should often only contain updates between relations (i.e. things that `cache.link` can take care of)

That said, this warning explicitly protects you from writing undefined values. Neither in development or production will Graphcache actually modify any values that are undefined in the data you provide it. This is simply to protect you because you've basically given it a fragment that doesn't make sense in that situation and it assumes you've made a mistake.

**Reply by rasmusolssony** [reply]
Thanks for the answer! I'm not really sure I understand what you mean though.

Let's say I have a query that looks like this:

```graphql
query {
   readers {
      id
      type
      port
      lastKey
      lastCard {
         cardData
         dateTime
         cardBitLength
      }
   }
}
```

I can get all data except for lastKey and lastCard from a fetch. I can then subscribe to get lastCard and lastKey when a card is presented or a key is pressed which only gives me that specific data.

How can I only update lastCard or lastKey without using writeFragment?
